### PR TITLE
OpenTelemetry instrumentation

### PR DESCRIPTION
**Pull request by ricklue**
In this pr, Ladybug is able to create telemetry data from Ladybug reports and send it to telemetry-collector Zipkin. Zipkin provides an overview based on time that helps to detect latency problems. To use Zipkin, you can run the following command: 
`docker run --rm -d -p 9411:9411 --name zipkin openzipkin/zipkin`
When a Ladybug report is created, a trace will now be sent to Zipkin.

Update: Ladybug is now also able to work with another tracing tool: Jaeger. To use Jaeger instead of Zipkin, you have to use the Jaeger exporter in the Config class instead of the Zipkin exporter and run the following command:
`docker run --rm -e COLLECTOR_ZIPKIN_HOST_PORT=:9411 -p 16686:16686 -p 4317:4317 -p 4318:4318 -p 9411:9411 jaegertracing/all-in-one:latest`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -308,4 +308,9 @@ When updating the front-end, work on a branch that is allowed to live on a fork
...
 
-At this point, you can merge your work on ladybug to its master branch. After this merge, additional pushes on your ladybug-frontend branch will succeed in CI/CD because the corresponding backend changes are in the backend master. You can merge your front-end changes to master if the GitHub actions test succeeds.
\ No newline at end of file
+At this point, you can merge your work on ladybug to its master branch. After this merge, additional pushes on your ladybug-frontend branch will succeed in CI/CD because the corresponding backend changes are in the backend master. You can merge your front-end changes to master if the GitHub actions test succeeds.
+
+OpenTelemetry
+=============
+
+Ladybug is able to create telemetry data from Ladybug reports and send it to telemetry-collector Zipkin. Zipkin provides an overview based on time that helps to detect latency problems. To use Zipkin, you can run the following command: `docker run --rm -d -p 9411:9411 --name zipkin openzipkin/zipkin` When a Ladybug report is created, a trace will now be sent to Zipkin.
```
Comment (jacodg) [inline README.md:316]: Please add info about Jaeger too (like you did in the conversation of this PR)

File: pom.xml [hunk 2]
```diff
@@ -52,6 +52,18 @@
     </developer>
   </developers>
 
+  <dependencyManagement>
```
Comment (jacodg) [inline pom.xml:55]: Why are you adding `<dependencyManagement>`? Can you remove it and only use `<dependencies>`?

File: src/main/java/nl/nn/testtool/Checkpoint.java [hunk 3]
```diff
@@ -104,6 +106,14 @@ public Checkpoint(Report report, String threadName, String sourceClassName,	Stri
 		this.level = level;
 	}
 
+	public Span getSpan() {
```
Comment (jacodg) [inline src/main/java/nl/nn/testtool/Checkpoint.java:109]: As this is a transient var I think you have to add @Transient and @JsonIgnore, see https://github.com/wearefrank/ladybug/blob/3440438628668b053ffe125e8cca0cdc6a95d630/src/main/java/nl/nn/testtool/Report.java#L79 The @JsonIgnore isn't documented I see but it is added to on most transient fields: https://github.com/wearefrank/ladybug/blob/3440438628668b053ffe125e8cca0cdc6a95d630/src/main/java/nl/nn/testtool/Report.java#L131

File: src/main/java/nl/nn/testtool/Checkpoint.java [hunk 4]
```diff
@@ -104,6 +106,14 @@ public Checkpoint(Report report, String threadName, String sourceClassName,	Stri
...
+		return span;
+	}
+
+	public void setSpan(Span span) {
```
Comment (jacodg) [inline src/main/java/nl/nn/testtool/Checkpoint.java:113]: Can you swap places of the set and get method (with all other set/get methods in this file the set comes before the get)?
Can you move the set and get method to the end of the file to keep the order of all the set/get methods in the file in sync with the order of the corresponding vars?

File: src/main/java/nl/nn/testtool/Report.java [hunk 5]
```diff
@@ -127,6 +133,14 @@ public class Report implements Serializable {
 	private transient boolean logMaxMemoryUsage = true;
 	private transient Map<Object, Set<Checkpoint>> streamingMessageListeners = new HashMap<>();
 	private transient Map<Object, StreamingMessageResult> streamingMessageResults = new HashMap<>();
+	private transient final OpenTelemetry openTelemetry;
+	private transient final Tracer tracer;
+
+	@Autowired
+	Report (OpenTelemetry openTelemetry) {
```
Comment (jacodg) [inline src/main/java/nl/nn/testtool/Report.java:140]: Do we need an extra constructor? Report has a reference to TestTool (var testTool), wouldn't it be better to do testTool.getOpenTelemetry() when needed?

File: src/main/java/nl/nn/testtool/TestTool.java [hunk 6]
```diff
@@ -94,6 +95,12 @@ public class TestTool {
 	private @Setter @Getter @Inject @Autowired Views views;
 	private @Setter @Getter int reportsInProgressThreshold = 300000;
 	boolean devMode = false; // See testConcurrentLastEndpointAndFirstStartpointForSameCorrelationId()
+	private final OpenTelemetry openTelemetry;
+
+	@Autowired
+	TestTool (OpenTelemetry openTelemetry) {
```
Comment (jacodg) [inline src/main/java/nl/nn/testtool/TestTool.java:101]: Why use a constructor? Can you change it to a setter?

File: src/main/java/nl/nn/testtool/Config.java [hunk 7]
```diff
@@ -101,6 +116,31 @@
 @Lazy // Lazy init singleton beans (prototype beans are already loaded on demand)
 @Configuration
 public class Config {
+	@Bean
+	OpenTelemetry openTelemetry() {
+		Resource resource = Resource.getDefault().toBuilder().put(ServiceAttributes.SERVICE_NAME, "ladybug").put(ServiceAttributes.SERVICE_VERSION, "1.0.0").build();
+		String endpointZipkin = "http://localhost:9411/api/v2/spans";
```
Comment (jacodg) [inline src/main/java/nl/nn/testtool/Config.java:122]: Shouldn't this be configurable? Does this code always enable Zipkin and Jaeger? Shouldn't it be configurable whether or not they should be enabled?

**Review by jacodg — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (ricklue) [reply README.md]: Done!

File: pom.xml (hunk 2 above)
Comment (ricklue) [reply pom.xml]: Added the change

File: src/main/java/nl/nn/testtool/Checkpoint.java (hunk 3 above)
Comment (ricklue) [reply src/main/java/nl/nn/testtool/Checkpoint.java]: Added the change

File: src/main/java/nl/nn/testtool/Checkpoint.java (hunk 4 above)
Comment (ricklue) [reply src/main/java/nl/nn/testtool/Checkpoint.java]: Added the change

File: src/main/java/nl/nn/testtool/Config.java (hunk 7 above)
Comment (ricklue) [reply src/main/java/nl/nn/testtool/Config.java:122]: I think I can't make this bean configurable, but I can make a condition in the other code to check if telemetry data have to be sent. If it's no, this bean will also not be used.

File: src/main/java/nl/nn/testtool/Report.java (hunk 5 above)
Comment (ricklue) [reply src/main/java/nl/nn/testtool/Report.java]: I removed this constructor

File: src/main/java/nl/nn/testtool/Config.java (hunk 7 above)
Comment (jacodg) [reply src/main/java/nl/nn/testtool/Config.java:122]: You can make the a separate bean of String endpointZipkin and check whether this bean is present or not to determine whether or not to use Zipkin and do something similar for Jaeger and have openTelemetry() return null when both are not configured. Wherever you use testTool.getOpenTelemetryTracer() you need to check whether it is null and do nothing when this is the case.

And then for example in the test webapp you can configure one of them in https://github.com/wearefrank/ladybug-test-webapp/blob/master/src/main/resources/springTestToolTestWebapp.xml by adding for example:

		<bean name="endpointZipkin" class="java.lang.String">
			<constructor-arg value="http://localhost:9411/api/v2/spans"/>
		</bean>

And something similar for Jaeger. Please add them in comment so pleople can enable them when needed and document it in the README.

**Review by jacodg — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 8]
```diff
@@ -245,6 +245,32 @@
...
+    <dependency>
+      <groupId>io.opentelemetry</groupId>
+      <artifactId>opentelemetry-exporter-zipkin</artifactId>
+      <version>1.39.0</version>
+    </dependency>
+    <dependency>
+      <groupId>io.opentelemetry</groupId>
+      <artifactId>opentelemetry-exporter-otlp</artifactId>
+      <version>1.39.0</version>
+    </dependency>
```
Comment (jacodg) [inline pom.xml:273]: Can you change all of them, except the first one, to provided? And add them without provided to the test webapp? I want to keep the number of dependencies that Ladybug adds to other applications to a minimum.

**Review by jacodg — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 8 above)
Comment (ricklue) [reply pom.xml:273]: Created pr https://github.com/wearefrank/ladybug-test-webapp/pull/22. I will change it to provided when this pr is merged

Comment (jacodg) [reply pom.xml:268]: Done

File: src/main/java/nl/nn/testtool/TestTool.java (hunk 6 above)
Comment (ricklue) [reply src/main/java/nl/nn/testtool/TestTool.java]: I removed the constructor

File: README.md [hunk 9]
```diff
@@ -243,4 +243,11 @@ What version number will be published by the Jenkins server? In the frontend `po
...
 > [!WARNING]  
-> When you run the Maven build on your development device, it will update `package.json`. Please do not check in that change. Otherwise, the build will not work for other developers anymore.
\ No newline at end of file
+> When you run the Maven build on your development device, it will update `package.json`. Please do not check in that change. Otherwise, the build will not work for other developers anymore.
+
+OpenTelemetry
+=============
+
+Ladybug is able to create telemetry data from Ladybug reports and send it to telemetry-collector Zipkin. Zipkin provides an overview based on time that helps to detect latency problems. To use Zipkin, you can run the following command: `docker run --rm -d -p 9411:9411 --name zipkin openzipkin/zipkin` When a Ladybug report is created, a trace will now be sent to Zipkin.
+
```
[the remark this answers is not in the document]
Comment (ricklue) [reply README.md]: Done

File: src/main/java/nl/nn/testtool/Report.java [hunk 10]
```diff
@@ -504,6 +510,17 @@ private  <T> T addCheckpoint(String threadName, String sourceClassName, String n
...
+		for (Checkpoint checkpointInList: checkpoints) {
+			if (checkpointInList.getType() == 1 && checkpointInList.getLevel() == checkpoint.getLevel() - 1) {
+				checkpointSpanBuilder.setParent(Context.current().with(checkpointInList.getSpan()));
+			}
+		}
+		Span checkpointSpan = checkpointSpanBuilder.startSpan();
+		checkpointSpan.setAttribute("checkpointType", checkpoint.getType());
+		checkpointSpan.setAttribute("checkpointTypeAsString", checkpoint.getTypeAsString());
+		checkpointSpan.setAttribute("checkpointLevel", checkpoint.getLevel());
+		checkpoint.setSpan(checkpointSpan);
```
[the remark this answers is not in the document]
Comment (ricklue) [reply src/main/java/nl/nn/testtool/Report.java]: I added an if statement

File: src/main/java/nl/nn/testtool/Report.java [hunk 11]
```diff
@@ -583,6 +600,17 @@ private  <T> T addCheckpoint(String threadName, String sourceClassName, String n
...
+			checkpoint.getSpan().end();
+		}
+		if (checkpointType == 2) {
+			for (Checkpoint checkpointInList: checkpoints) {
+				if (checkpointInList.getType() == 1 && checkpointInList.getLevel() == checkpoint.getLevel() - 1) {
+					checkpointInList.getSpan().end();
+				}
+			}
+		}
+
```
[the remark this answers is not in the document]
Comment (ricklue) [reply src/main/java/nl/nn/testtool/Report.java:617]: Added the change

**Review by jacodg — APPROVED** [review]
